src/lc3/vm.rs

```diff
@@ -0,0 +1,806 @@
+use std::io::{stdin, stdout, Read, Write};
+
+use super::{bytes::sign_extend, memory::Memory, opcode::Opcode, trap::TrapCode};
+
+const TOTAL_REGISTERS: usize = 8;
+
+pub struct VM {
+    pub registers: [u16; TOTAL_REGISTERS],
+    pub pc: u16,
+    pub cond: ConditionFlag,
+    pub running: bool,
+    pub memory: Memory,
+}
+
+#[derive(Debug)]
+pub enum VMError {
+    ReadingFile(String),
+    ConcatenatingBytes(String),
+    Overflow,
+    MemoryIndex(String),
+    InvalidOpcode,
+    InvalidRegister,
+    InvalidTrapCode,
+    StandardIO(String),
+    InvalidCharacter,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ConditionFlag {
+    Pos,
+    Zro,
+    Neg,
+}
+
+impl VM {
+    pub fn new() -> VM {
+        VM {
+            registers: [0; TOTAL_REGISTERS],
+            pc: 0x3000,
+            cond: ConditionFlag::Zro,
+            running: false,
+            memory: Memory::new(),
+        }
+    }
+
+    /// Get value stored in the register requested
+    ///
+    /// If register is not 0 to 7, an Error is returned
```

Comment:
I would include `register_index` to document how the function argument affects the function logic. For example:

```suggestion
    /// Get value stored in the register requested in `register_index`
    ///
    /// If `register_index` is outside of the range from 0 to 7, an Error is returned
```

Reply:
Done, [2cad29e](https://github.com/tomip01/lc3-vm/pull/6/commits/2cad29ef1cfb80e565137d58bbcd328bcfc50b06)